storage/Devices/BlkDeviceImpl.cc

```diff
@@ -208,9 +209,42 @@ namespace storage
     }
 
 
+    bool
+    BlkDevice::Impl::exists_by_name(const Devicegraph* devicegraph, const string& name,
+				    SystemInfo& system_info)
+    {
+	if (!devicegraph->get_impl().is_probed())
+	    ST_THROW(Exception("function called on wrong devicegraph"));
+
+	for (Devicegraph::Impl::vertex_descriptor vertex : devicegraph->get_impl().vertices())
+	{
+	    const BlkDevice* blk_device = dynamic_cast<const BlkDevice*>(devicegraph->get_impl()[vertex]);
+	    if (blk_device)
+	    {
+		if (blk_device->get_name() == name)
+		    return true;
+	    }
+	}
+
+	dev_t majorminor = system_info.getCmdUdevadmInfo(name).get_majorminor();
+
+	for (Devicegraph::Impl::vertex_descriptor vertex : devicegraph->get_impl().vertices())
+	{
+	    const BlkDevice* blk_device = dynamic_cast<const BlkDevice*>(devicegraph->get_impl()[vertex]);
+	    if (blk_device && blk_device->get_impl().active)
+	    {
+		if (system_info.getCmdUdevadmInfo(blk_device->get_name()).get_majorminor() == majorminor)
+		    return true;
+	    }
+	}
+
```

Comment:
Yes, the same devices are interated. But the first iteration only does a match for the name in memory while the second does a real system lookup. The idea is to avoid the system lookup if the first iteration find a match.